src/expr/src/scalar/mod.rs

```diff
@@ -1961,6 +1982,129 @@ impl VisitChildren<Self> for MirScalarExpr {
     }
 }
 
+/// Filter characteristics that are used for ordering join inputs.
+/// This can be created for a `Vec<MirScalarExpr>`, which represents an AND of predicates.
+///
+/// The fields should be ordered by typical selectivity, so that Ord gives the right
+/// ordering for join inputs. Bigger is better, i.e., will tend to come to the front of the inputs.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Debug, Clone, Serialize, Deserialize, Hash, MzReflect)]
+pub struct FilterCharacteristics {
+    literal_equality: bool,
+    // (Assuming a random string of lower-case characters, `LIKE 'a%'` has a selectivity of 1/26.)
+    like: bool,
+    is_null: bool,
+    // Number of Vec elements that involve inequality predicates.
+    // A BETWEEN is represented as two inequality predicates.
+    // (For inequality predicates, some databases assume 1/3 selectivity in the absence of statistics.)
+    literal_inequality: usize,
+    /// Any filter, except ones involving `IS NOT NULL`, because those are too common.
+    /// Can be true by itself, or any other field being true can also make this true.
+    /// `NOT LIKE` is only in this category.
+    /// `!=` is only in this category.
+    /// `NOT (a = b)` is turned into `!=` by `reduce` before us!
+    any_filter: bool,
+}
+
+impl BitOrAssign for FilterCharacteristics {
+    fn bitor_assign(&mut self, rhs: Self) {
+        self.literal_equality |= rhs.literal_equality;
+        self.like |= rhs.like;
+        self.is_null |= rhs.is_null;
+        self.literal_inequality += rhs.literal_inequality;
+        self.any_filter |= rhs.any_filter;
+    }
+}
+
+impl FilterCharacteristics {
+    pub fn none() -> FilterCharacteristics {
+        FilterCharacteristics {
+            literal_equality: false,
+            like: false,
+            is_null: false,
+            literal_inequality: 0,
+            any_filter: false,
+        }
+    }
+
+    pub fn filter_characteristics(
+        filters: &Vec<MirScalarExpr>,
+    ) -> Result<FilterCharacteristics, RecursionLimitError> {
+        let mut literal_equality = false;
+        let mut like = false;
+        let mut is_null = false;
+        let mut literal_inequality = 0;
+        let mut any_filter = false;
+        filters.iter().try_for_each(|f| {
+            let mut literal_inequality_in_current_filter = false;
+            let mut is_not_null_in_current_filter = false;
+            f.visit_pre_with_context(
+                false,
+                &mut |not_in_parent_chain, expr| {
+                    not_in_parent_chain
+                        || matches!(
+                            expr,
+                            MirScalarExpr::CallUnary {
+                                func: UnaryFunc::Not(func::Not),
+                                ..
+                            }
+                        )
+                },
+                &mut |not_in_parent_chain, expr| {
+                    if !not_in_parent_chain {
```

Comment:
Can you put a comment here explaining why you guard the logic below in `!not_in_parent_chain` for posteriy (it's not immediately obvious).